src/server/editorServices.ts

```diff
@@ -0,0 +1,1646 @@
+        errorMsg?: string;
+        project?: Project;
+    }
+
+    function copyListRemovingItem<T>(item: T, list: T[]) {
+        var copiedList: T[] = [];
+        for (var i = 0, len = list.length; i < len; i++) {
+            if (list[i] != item) {
+                copiedList.push(list[i]);
+            }
+        }
+        return copiedList;
+    }
+
+    interface ProjectServiceEventHandler {
+        (eventName: string, project: Project): void;
+    }
+
+    export class ProjectService {
+        filenameToScriptInfo: ts.Map<ScriptInfo> = {};
+        // open, non-configured files in two lists
+        openFileRoots: ScriptInfo[] = [];
+        openFilesReferenced: ScriptInfo[] = [];
+        // projects covering open files
+        inferredProjects: Project[] = [];
+
+        constructor(public host: ServerHost, public psLogger: Logger, public eventHandler?: ProjectServiceEventHandler) {
+            // ts.disableIncrementalParsing = true;
+        }
+
+        watchedFileChanged(fileName: string) {
+            var info = this.filenameToScriptInfo[fileName];
+            if (!info) {
+                this.psLogger.info("Error: got watch notification for unknown file: " + fileName);
+            }
+
+            if (!this.host.fileExists(fileName)) {
+                // File was deleted
+                this.fileDeletedInFilesystem(info);
+            }
+            else {
+                if (info && (!info.isOpen)) {
+                    info.svc.reloadFromFile(info.fileName);
+                }
+            }
+        }
+        
+
+        log(msg: string, type = "Err") {
+            this.psLogger.msg(msg, type);
+        }
+
+        closeLog() {
+            this.psLogger.close();
+        }
+
+        createInferredProject(root: ScriptInfo) {
+            var iproj = new Project(this);
+            iproj.addRoot(root);
+            iproj.finishGraph();
+            this.inferredProjects.push(iproj);
+            return iproj;
+        }
+
+        fileDeletedInFilesystem(info: ScriptInfo) {
+            this.psLogger.info(info.fileName + " deleted");
+            
+            if (info.fileWatcher) {
+                info.fileWatcher.close();
+                info.fileWatcher = undefined;
+            }
+
+            if (!info.isOpen) {
+                this.filenameToScriptInfo[info.fileName] = undefined;
+                var referencingProjects = this.findReferencingProjects(info);
+                for (var i = 0, len = referencingProjects.length; i < len; i++) {
+                    referencingProjects[i].removeReferencedFile(info);
+                }
+            }
+            this.printProjects();
+        }
+
+        addOpenFile(info: ScriptInfo) {
+            this.findReferencingProjects(info);
+            if (info.defaultProject) {
+                this.openFilesReferenced.push(info);
+            }
+            else {
+                // create new inferred project p with the newly opened file as root

```

Comment:
Maybe give a big-picture comment rather than a step-by-step comment for what's happening here.